subm/src/subm.js

```diff
@@ -427,23 +521,48 @@ async function main(env, { clock, get, express, makeStorage, admin }) {
       error: app.get('env') === 'development' ? err : {},
     });
   };
-  // Upload form
-  // Note the actual upload request goes directly to Google Cloud Storage.
-  app.get(Site.path.uploadSlog, loginCheck, async (req, res) => {
-    try {
-      const participant =
-        /** @type { ReturnType<typeof makeTestnetParticipant> } */ (req.user);
-      const policy = await participant.uploadPolicy(
-        new Date(clock()),
-        Site.uploadSuccessURL(base, req.protocol, req.hostname),
-      );
-      const files = await participant.myFiles();
-      const page = Site.uploadSlog(participant.member, policy, files);
-      res.send(page);
-    } catch (err) {
-      handleError(res, req.baseUrl, err);
+
+  app.get(Site.path.contactForm, loginCheck, async (req, res) => {
+    const contact = /** @type { ReturnType<typeof makeContact> } */ (req.user);
+    const { user } = contact.member;
+    if (!user) throw Error('no user');
+    const detail = await discordAPI.users(user.id);
+    let email = detail.email; // ISSUE: why is email missing? what bot permission am I missing?
+    if (!email) {
+      email = await contact.getEmail();
     }
+    const page = Site.contactForm(contact.member, email, req.query.ack === '1');
+    res.send(page);
+  });
+  app.post(Site.path.contactForm, loginCheck, async (req, res) => {
+    const contact = /** @type { ReturnType<typeof makeContact> } */ (req.user);
+    const { email } = req.body;
```

Comment:
I assume body parsing is done safely by the framework?

Reply:
yes; see `app.use(express.urlencoded({ extended: true }));` above.